Kafka stream

ksinghbora opened the pull request:
This is a new implementation of Kafka-connector which leverage the power of kstream and ktable. This is the natural way of table query because of ktable implementation.

This commit include :
- Initial end to end working code for Kafka using kstream and ktable.
- Json parser to parse json content from kafka and read as a field.
- Sample metadata.json.

My future task:
- Full wiki page (will come soon)
- Multiple parser support for kafka content
- Configurable Ktable storage (Global KTable, Local storage, centralized storage)
- Test cases (parallel task)

kokosing commented on presto-kafka-stream/HEADER.txt:
```diff
@@ -0,0 +1,13 @@
+/*
```
Please remove this file

ksinghbora replied:
Done

kokosing commented on .gitignore:
```diff
@@ -24,3 +24,4 @@ benchmark_outputs
 .checkstyle
 .mvn/timing.properties
 .editorconfig
+.metadata/*
```
What creates these files?

ksinghbora replied:
I am getting file like .metadata/.lock

kokosing commented on presto-kafka-stream/pom.xml:
```diff
@@ -0,0 +1,133 @@
+
```
missing xml header

ksinghbora replied:
added

kokosing commented on presto-kafka-stream/kafkastream.properties:
```diff
@@ -0,0 +1,3 @@
+connector.name=kafkastream
```
What is the purpose of this file here?

ksinghbora replied:
I didn't add this in presto-main, so for now I am keeping this here as an example. once connector is included, will remove from here.

kokosing commented on presto-kafka-stream/src/main/java/com/facebook/presto/kafkastream/JsonFlattener.java:
```diff
@@ -0,0 +1,612 @@
+ * Class is used to flatten the json data. This will only select the values
+ * specified by jsonPaths.
+ */
+public final class JsonFlattener
```
can you please write unit tests for this class?

ksinghbora replied:
Test cases is already written but in Junit. I will change this to ngTest. Will commit once done.

kokosing commented:
- Why this was not added as an extension of existing kafka connector?
- There are no tests.

Please ping me once you write documentation for this.

findepi commented:
It's unlikely we want to maintain two kafka connectors within Presto code base. If this is strictly better, then we should amend/expand/refactor the existing connector with your code. All existing tests should stay and new ones should be added to cover this "being better".

ksinghbora commented:
> It's unlikely we want to maintain two kafka connectors within Presto code base. If this is strictly better, then we should amend/expand/refactor the existing connector with your code. All existing tests should stay and new ones should be added to cover this "being better".

ksinghbora commented:
@kokosing , I added initial documentation. Please go through it. Let me know If any change required.

ksinghbora commented:
> It's unlikely we want to maintain two kafka connectors within Presto code base. If this is strictly better, then we should amend/expand/refactor the existing connector with your code. All existing tests should stay and new ones should be added to cover this "being better".

I wrote this connector because of better compatibility between KTable and presto data-store concept. Reason to create new connector is, I am using different kafka client than current one which have couple of new terminology. 
This connector is initial release , I am planning to optimize further with multiple option.
Most of the test cases has been written already in junit. I am converting it to NgTest.
Please let me know your plan for this. I will resume working on, once you finalize.

sopel39 commented:
> Reason to create new connector is, I am using different kafka client than current one which have couple of new terminology.

What do you mean by different kafka client?

ksinghbora commented:
> > Reason to create new connector is, I am using different kafka client than current one which have couple of new terminology.
> 
> What do you mean by different kafka client?

I am using Kafka-streams client:

<dependency>
			<groupId>org.apache.kafka</groupId>
			<artifactId>kafka-streams</artifactId>
			<version>1.0.0</version>
		</dependency>
Apache kafka has introduced this sometime back.

ksinghbora commented:
I understand that there is no mean to have two connector for the same purpose and we should maintain single kafka connector eventually. 
In order to merge we have to made some decision and finalise design.

Some of the decision points are:
1 - Support for both Kafka Client ?
2- Design for merge.

I can come up with design proposal. Please let me know If there is any specific format?

findepi commented:
> 1 - Support for both Kafka Client ?

@ksinghbora i'm no Kafka expert, so forgive this silly question: what would be the benefits of supporting both Kafka clients?

>  Please let me know If there is any specific format?

You can for example share a google doc. This would allow to comment on the document inline.

ksinghbora commented:
> > 1 - Support for both Kafka Client ?
> 
> @ksinghbora i'm no Kafka expert, so forgive this silly question: what would be the benefits of supporting both Kafka clients?
> 
> > Please let me know If there is any specific format?
> 
> You can for example share a google doc. This would allow to comment on the document inline.

I have gone through the current kafka connector implementation on high level. As per my understanding, we are retrieving all offset and then data for each request. We are scanning whole data for each request which cause performance impact.(correct me if I am wrong).

New connector provide the index capability on kafka key (We can index other fields also, not implemented yet). Only non index fields will require full scan. Overall it makes connector performant.

@findepi  We don't really require both client. However implementation wise both connectors are different. Merging will remove one completely.

ksinghbora commented:
Any Conclusion on this ?

ksinghbora commented:
Are we still open for Kafka Connector enhancement?

On Thu, Apr 4, 2019 at 2:39 PM stale[bot] <notifications@github.com> wrote:

> This pull request has been automatically marked as stale because it has
> not had recent activity. If you'd still like this PR merged, please comment
> on the task, make sure you've addressed reviewer comments, and rebase on
> the latest master. Thank you for your contributions!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/11609#issuecomment-479815885>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrusVknpa3v62BpJbjiqNOY-0QpgxAIks5vdcFHgaJpZM4XAhKl>
> .
>


-- 
with regards
-----------------------
Kundan Singh Bora
